### PR TITLE
[local_auth] updated getAvailableBiometrics behavior on Android

### PR DESCRIPTION
**Pull request by darwin-morocho**
On Android `canCheckBiometrics` returns true if the device doesn't have any biometric enrolled (For more info check https://github.com/flutter/flutter/issues/45497#issuecomment-1007310214). Now with this PR `canCheckBiometrics` returns false on Android if device has the biometric hardware but doesn't have any biometric enrolled.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by carman247** [comment]
following

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java [hunk 1]
```diff
@@ -61,4 +62,32 @@ public void onDetachedFromActivity_ShouldReleaseActivity() {
     plugin.onDetachedFromActivity();
     assertNull(plugin.getActivity());
   }
+
+  @Test
+  public void getAvailableBiometrics_CanCheckBiometricsReturnsFalse() {
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java:67]: This name is unrelated to what the test is asserting.

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java [hunk 2]
```diff
@@ -61,4 +62,32 @@ public void onDetachedFromActivity_ShouldReleaseActivity() {
...
+    when(mockLifecycleReference.getLifecycle()).thenReturn(mockLifecycle);
+    when(mockPluginBinding.getFlutterEngine()).thenReturn(mockFlutterEngine);
+    when(mockFlutterEngine.getDartExecutor()).thenReturn(mockDartExecutor);
+
+    final LocalAuthPlugin plugin = new LocalAuthPlugin();
+    plugin.onAttachedToEngine(mockPluginBinding);
+    plugin.onAttachedToActivity(mockActivityBinding);
+    plugin.onMethodCall(new MethodCall("getAvailableBiometrics", null), mockResult);
+    verify(mockResult).success(Collections.emptyList());
+  }
```
[the remark this answers is not in the document]
Comment (darwin-morocho) [reply packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java:101]: done 🤓

File: packages/local_auth/local_auth/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## 1.2.0
```
[the remark this answers is not in the document]
Comment (darwin-morocho) [reply packages/local_auth/local_auth/CHANGELOG.md]: changed

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java (hunk 1 above)
Comment (darwin-morocho) [reply packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java]: fixed and improved

File: packages/local_auth/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java [hunk 4]
```diff
@@ -257,6 +258,10 @@ private void getAvailableBiometrics(final Result result) {
         result.error("no_activity", "local_auth plugin requires a foreground activity", null);
         return;
       }
+      if (!canAuthenticateWithBiometrics()) {
+        result.success(Collections.emptyList());
```
[the remark this answers is not in the document]
Comment (darwin-morocho) [reply packages/local_auth/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java]: done

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java [hunk 5]
```diff
@@ -20,45 +24,95 @@
...
+  @Mock private DartExecutor mockDartExecutor;
+  @Mock private ActivityPluginBinding mockActivityBinding;
+  @Mock private MethodChannel.Result mockResult;
+  @Mock private FlutterPluginBinding mockPluginBinding;
 
-    Context mockContext = mock(Context.class);
-    when(mockActivity.getBaseContext()).thenReturn(mockContext);
+  @Before
+  public void setUp() {
+    plugin = new LocalAuthPlugin();
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java:48]: Do not make the object under test part of the test fixture. It's an anti-pattern that makes understanding tests much more difficult.

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java [hunk 6]
```diff
@@ -20,45 +24,95 @@
...
     plugin.onAttachedToEngine(mockPluginBinding);
     plugin.onAttachedToActivity(mockActivityBinding);
     assertNotNull(plugin.getActivity());
 
     plugin.onDetachedFromActivity();
     assertNull(plugin.getActivity());
   }
+
+  @Test
+  public void getAvailableBiometrics_ShouldReturnUndefinedBiometrics() {
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java:89]: These names should describe the condition as well as the result, otherwise they are hard to follow; as written the tests sound like they are asserting three different behaviors for the same situation (which makes no sense). E.g., `getAvailableBiometrics_withHardwareButNothingEnrolled_ShouldReturnUndefined`

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java [hunk 7]
```diff
@@ -20,45 +24,95 @@
...
+    plugin.onMethodCall(new MethodCall("getAvailableBiometrics", null), mockResult);
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java:90]: You should be explicitly setting the intended preconditions; it is not at all clear what conditionals you are actually testing here. You should be testing `BIOMETRIC_ERROR_NONE_ENROLLED` for the case of nothing being enrolled.

File: packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java [hunk 8]
```diff
@@ -20,45 +24,95 @@
...
+    verify(mockResult).success(Collections.singletonList("undefined"));
+  }
+
+  @Test
+  public void getAvailableBiometrics_ShouldReturnEmptyBiometrics() {
+    final BiometricManager mockBiometricManager = mock(BiometricManager.class);
+    when(mockBiometricManager.canAuthenticate()).thenReturn(BiometricManager.BIOMETRIC_SUCCESS);
+    plugin.overrideBiometricManager(mockBiometricManager);
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/android/src/test/java/io/flutter/plugins/localauth/LocalAuthTest.java:98]: The preconditions here are unclear as well. Why does `SUCCESS` correspond to nothing? You should be explicitly returning false for the features.

But also, this test is not actually a valid case. The docs say "Returns BIOMETRIC_SUCCESS if a strong biometric can currently be used (enrolled and available)." So you are actually asserting the wrong behavior here, because you've created an invalid scenario, where the mock APIs are giving conflicting answers on enrollment.

File: packages/local_auth/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java [hunk 9]
```diff
@@ -269,13 +271,19 @@ private void getAvailableBiometrics(final Result result) {
     if (activity == null || activity.isFinishing()) {
       return biometrics;
     }
+
+    if (!canAuthenticateWithBiometrics()) {
+      biometrics.add("undefined");
```
Comment (stuartmorgan) [inline packages/local_auth/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:276]: You are returning "undefined" even on a device that has no hardware, which is wrong.

That means you are also missing critical test cases.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by RafaRuiz** [comment]
Why was this closed?
@darwin-morocho 
@stuartmorgan

**Comment by darwin-morocho** [comment]
> Why was this closed? @darwin-morocho @stuartmorgan
> 
> I can finish the requested changes if needed.

@RafaRuiz I closed this due to At the moment I don't have much time to solve the all changes. I reopen and you can finish the requested changes

**Comment by RafaRuiz** [comment]
@darwin-morocho I fork it from yours, fix the requested changes from @stuartmorgan and open a new PR (as I can't participate in this one)
